pkg/object/ingresscontroller/translator.go

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ingresscontroller
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/object/httpserver"
+	"github.com/megaease/easegress/pkg/supervisor"
+	apicorev1 "k8s.io/api/core/v1"
+	apinetv1 "k8s.io/api/networking/v1"
+)
+
+const (
+	defaultPipelineName = "pipeline-default"
+)
+
+// specTranslator translates k8s ingress related specs to Easegress http server
+// spec and pipeline specs
+type specTranslator struct {
+	k8sClient    *k8sClient
+	httpSvr      *supervisor.Spec
+	pipelines    map[string]*supervisor.Spec
+	httpSvrCfg   *httpserver.Spec
+	ingressClass string
+}
+
+func newSpecTranslator(k8sClient *k8sClient, ingressClass string, httpSvrCfg *httpserver.Spec) *specTranslator {
+	return &specTranslator{
+		k8sClient:    k8sClient,
+		httpSvrCfg:   httpSvrCfg,
+		ingressClass: ingressClass,
+	}
+}
+
+func (st *specTranslator) httpServerSpec() *supervisor.Spec {
+	return st.httpSvr
+}
+
+func (st *specTranslator) pipelineSpecs() map[string]*supervisor.Spec {
+	return st.pipelines
+}
+
+func generatePipelineSpec(name string, endpoints []string) (*supervisor.Spec, error) {
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "name: %s\n", name)
+	buf.WriteString(`kind: HTTPPipeline
+flow:
```

Comment:
According to @xxx7xxxx 's advice in MeshController and FaaSController, proposal to use structure builder[1]

## Reference
1. https://github.com/megaease/easegress/blob/main/pkg/object/function/worker/ingress.go#L62